selfservice/strategy/passkey/passkey_registration.go

```diff
@@ -0,0 +1,358 @@
+
+package passkey
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/go-webauthn/webauthn/protocol"
+	"github.com/go-webauthn/webauthn/webauthn"
+	"github.com/pkg/errors"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+
+	"github.com/ory/herodot"
+	jsonschema "github.com/ory/jsonschema/v3"
+	"github.com/ory/kratos/identity"
+	"github.com/ory/kratos/schema"
+	"github.com/ory/kratos/selfservice/flow"
+	"github.com/ory/kratos/selfservice/flow/registration"
+	"github.com/ory/kratos/text"
+	"github.com/ory/kratos/ui/container"
+	"github.com/ory/kratos/ui/node"
+	"github.com/ory/kratos/x"
+	"github.com/ory/kratos/x/webauthnx"
+	"github.com/ory/x/randx"
+)
+
+//go:embed .schema/registration.schema.json
+var registrationSchema []byte
+
+func (s *Strategy) RegisterRegistrationRoutes(r *x.RouterPublic) {
+	webauthnx.RegisterWebauthnRoute(r)
+}
+
+func (s *Strategy) PopulateRegistrationMethod(r *http.Request, regFlow *registration.Flow) error {
+	ctx := r.Context()
+
+	if regFlow.Type != flow.TypeBrowser {
+		return nil
+	}
+
+	defaultSchemaURL, err := s.d.Config().DefaultIdentityTraitsSchemaURL(ctx)
+	if err != nil {
+		return err
+	}
+	nodes, err := s.registrationNodes(ctx, defaultSchemaURL)
+	if err != nil {
+		return err
+	}
+
+	for _, n := range nodes {
+		regFlow.UI.SetNode(n)
+	}
+
+	regFlow.UI.Nodes.Append(node.NewInputField(
+		"method",
+		"passkey",
+		node.PasskeyGroup,
+		node.InputAttributeTypeSubmit,
+	).WithMetaLabel(text.NewInfoSelfServiceRegistrationRegisterPasskey()))
+
+	regFlow.UI.SetCSRF(s.d.GenerateCSRFToken(r))
+
+	return nil
+}
+
+func (s *Strategy) registrationNodes(ctx context.Context, schemaURL *url.URL) (node.Nodes, error) {
+	runner, err := schema.NewExtensionRunner(ctx)
+	if err != nil {
+		return nil, err
+	}
+	c := jsonschema.NewCompiler()
+	runner.Register(c)
+
+	nodes, err := container.NodesFromJSONSchema(ctx, node.DefaultGroup, schemaURL.String(), "", c)
+	if err != nil {
+		return nil, err
+	}
+	return nodes, nil
+}
+
+// webauthnIdentifierNode returns the node that is used to identify the user in the WebAuthn flow.
+func (s *Strategy) webauthnIdentifierNode(ctx context.Context, schemaURL *url.URL) (*node.Node, error) {
+	nodes, err := s.registrationNodes(ctx, schemaURL)
+	if err != nil {
+		return nil, err
+	}
+	for _, n := range nodes {
+		if attr, ok := n.Attributes.(*node.InputAttributes); ok {
+			if attr.DataWebauthnIdentifier {
+				return n, nil
+			}
+		}
+	}
+
+	return nil, schema.NewMissingIdentifierError()
+}
+
+func (s *Strategy) Register(w http.ResponseWriter, r *http.Request, regFlow *registration.Flow, ident *identity.Identity) (err error) {
+	ctx := r.Context()
+
+	if regFlow.Type != flow.TypeBrowser {
+		return flow.ErrStrategyNotResponsible
+	}
+
+	params, err := s.decode(r)
+	if err != nil {
+		return s.handleRegistrationError(w, r, regFlow, params, err)
+	}
+	if params.Register == "" && params.Method != "passkey" {
```

Comment:
Not in a benign case, because the register param is set right before submission.

As for an attacker, what would the gain be?

Reply:
Requiring the method to be set will further complicate submitting the passkey, which is done through a form submit in JS, see: https://github.com/ory/kratos/blob/96c2164e2ad5befa1b05be14b6011556b88ec136/x/webauthnx/js/webauthn.js#L121

We don't require the method to be set for webauthn register requests either, btw: https://github.com/ory/kratos/blob/96c2164e2ad5befa1b05be14b6011556b88ec136/selfservice/strategy/webauthn/registration.go#L94-L114

Reply:
Specifically, I could add a hidden input with name=method and value=passkey in JS. Is it worth it?

Reply:
I'd say it's worth the effort also submitting the method, since we do have the (sometimes loose) convention to submit the method as well. Since this would just be another edge case, potentially causing issues down the road.

But, in the interest of velocity, I'll leave the decision up to you.